File: components/mardown-display/blocks/transcripts/transcript-parser.ts
import { TranscriptSegment } from "./AdvancedTranscriptViewer";

export const parseTranscriptContent = (transcriptContent: string): TranscriptSegment[] => {
    const lines = transcriptContent.split("\n");

    let segments: TranscriptSegment[] = [];
    let currentSegment: Partial<TranscriptSegment> = {};
    let buffer = "";

    for (let i = 0; i < lines.length; i++) {
        const line = lines[i].trim();

        // Skip empty lines or title lines
        if (line === "" || line.startsWith("**Audio Transcription**")) continue;

        // Check if line contains a timecode [MM:SS] or [HH:MM:SS]
        const timecodeMatch = line.match(/^\[(\d+):(\d+)(?::(\d+))?\]/);

        if (timecodeMatch) {
            // If we have accumulated text, save the previous segment
            if (buffer.trim() && currentSegment.timecode) {
                segments.push({
                    id: `segment-${segments.length}`,
                    timecode: currentSegment.timecode || "",
                    seconds: currentSegment.seconds || 0,
                    text: buffer.trim(),
                    speaker: currentSegment.speaker,
                });
                buffer = "";
            }

            // Parse the timecode - handle both MM:SS and HH:MM:SS formats
            let hours = 0;
            let minutes = 0;
            let seconds = 0;

            if (timecodeMatch[3]) {
                // HH:MM:SS format
                hours = parseInt(timecodeMatch[1], 10);
                minutes = parseInt(timecodeMatch[2], 10);
                seconds = parseInt(timecodeMatch[3], 10);
            } else {
                // MM:SS format
                minutes = parseInt(timecodeMatch[1], 10);
                seconds = parseInt(timecodeMatch[2], 10);
            }

            const totalSeconds = hours * 3600 + minutes * 60 + seconds;

            // Format timecode string based on presence of hours
            let timecodeStr = "";
            if (hours > 0) {
                timecodeStr = `${hours.toString().padStart(2, "0")}:${minutes.toString().padStart(2, "0")}:${seconds
                    .toString()
                    .padStart(2, "0")}`;
            } else {
                timecodeStr = `${minutes.toString().padStart(2, "0")}:${seconds.toString().padStart(2, "0")}`;
            }

            // Start a new segment
            currentSegment = {
                timecode: timecodeStr,
                seconds: totalSeconds,
            };

            // Extract any speaker info or additional text on the same line
            const restOfLine = line.replace(timecodeMatch[0], "").trim();

            if (restOfLine) {
                // Check for speaker pattern like "Speaker A:" or similar
                const speakerMatch = restOfLine.match(/^([^:]+):\s*/);

                // Also check for bold format like **Speaker**
                const boldSpeakerMatch = restOfLine.match(/^(\*\*[^*]+\*\*)/);

                if (speakerMatch) {
                    currentSegment.speaker = speakerMatch[1].trim();
                    buffer += restOfLine.substring(speakerMatch[0].length) + " ";
                } else if (boldSpeakerMatch) {
                    currentSegment.speaker = boldSpeakerMatch[1].replace(/\*/g, "").trim();
                    buffer += restOfLine.replace(boldSpeakerMatch[0], "") + " ";
                } else {
                    buffer += restOfLine + " ";
                }
            }
        } else {
            // Continue with the current segment
            buffer += line + " ";
        }
    }

    // Don't forget the last segment
    if (buffer.trim() && currentSegment.timecode) {
        segments.push({
            id: `segment-${segments.length}`,
            timecode: currentSegment.timecode || "",
            seconds: currentSegment.seconds || 0,
            text: buffer.trim(),
            speaker: currentSegment.speaker,
        });
    }

    return segments;
};

File: lib/python/parsers/transcript_parser.py
"""Transcript parser — port of transcript-parser.ts."""

from __future__ import annotations

import re

from lib.python.models.transcript import TranscriptBlockData, TranscriptSegment

_TIMECODE_RE = re.compile(r'^\[(\d+):(\d+)(?::(\d+))?\]')
_SPEAKER_RE = re.compile(r'^([^:]+):\s*')
_BOLD_SPEAKER_RE = re.compile(r'^(\*\*[^*]+\*\*)')


def parse_transcript(content: str) -> TranscriptBlockData:
    """Parse timecoded transcript content."""
    lines = content.split("\n")
    segments: list[TranscriptSegment] = []
    current_timecode: str | None = None
    current_seconds: float = 0
    current_speaker: str | None = None
    buffer = ""

    for line in lines:
        stripped = line.strip()
        if not stripped or stripped.startswith("**Audio Transcription**"):
            continue

        tc_match = _TIMECODE_RE.match(stripped)
        if tc_match:
            # Save previous segment
            if buffer.strip() and current_timecode is not None:
                segments.append(TranscriptSegment(
                    id=f"segment-{len(segments)}",
                    timecode=current_timecode,
                    seconds=current_seconds,
                    text=buffer.strip(),
                    speaker=current_speaker,
                ))
                buffer = ""

            # Parse timecode
            if tc_match.group(3):
                hours = int(tc_match.group(1))
                minutes = int(tc_match.group(2))
                seconds = int(tc_match.group(3))
            else:
                hours = 0
                minutes = int(tc_match.group(1))
                seconds = int(tc_match.group(2))

            total_seconds = hours * 3600 + minutes * 60 + seconds

            if hours > 0:
                tc_str = f"{hours:02d}:{minutes:02d}:{seconds:02d}"
            else:
                tc_str = f"{minutes:02d}:{seconds:02d}"

            current_timecode = tc_str
            current_seconds = float(total_seconds)
            current_speaker = None

            rest = stripped[tc_match.end():].strip()
            if rest:
                speaker_m = _SPEAKER_RE.match(rest)
                bold_m = _BOLD_SPEAKER_RE.match(rest)
                if speaker_m:
                    current_speaker = speaker_m.group(1).strip()
                    buffer += rest[speaker_m.end():] + " "
                elif bold_m:
                    current_speaker = bold_m.group(1).replace("*", "").strip()
                    buffer += rest[bold_m.end():] + " "
                else:
                    buffer += rest + " "
        else:
            buffer += stripped + " "

    # Last segment
    if buffer.strip() and current_timecode is not None:
        segments.append(TranscriptSegment(
            id=f"segment-{len(segments)}",
            timecode=current_timecode,
            seconds=current_seconds,
            text=buffer.strip(),
            speaker=current_speaker,
        ))

    return TranscriptBlockData(segments=segments)
